noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -1,76 +1,121 @@
-mod lib;
-mod settings;
+mod config;
 
 use dep::aztec::macros::aztec;
 
+/// Fee Payment Contract (FPC) allows users to pay for the transaction fee with an arbitrary asset. Supports private
+/// and public fee payment flows.
 #[aztec]
 contract FPC {
-    use crate::lib::compute_rebate;
-    use crate::settings::Settings;
+    use crate::config::Config;
     use dep::aztec::{
-        macros::{functions::{initializer, internal, private, public}, storage::storage},
+        macros::{functions::{initializer, internal, private, public, view}, storage::storage},
         protocol_types::{abis::function_selector::FunctionSelector, address::AztecAddress},
         state_vars::PublicImmutable,
     };
     use dep::token::Token;
 
     #[storage]
     struct Storage<Context> {
-        settings: PublicImmutable<Settings, Context>,
+        config: PublicImmutable<Config, Context>,
     }
 
+    /// Initializes the contract with an accepted asset (AA) and a private fee recipient (address that receives AA
+    /// from users during the private fee payment flow).
     #[public]
     #[initializer]
-    fn constructor(other_asset: AztecAddress, admin: AztecAddress) {
-        let settings = Settings { other_asset, admin };
-        storage.settings.initialize(settings);
+    fn constructor(accepted_asset: AztecAddress, private_fee_recipient: AztecAddress) {
+        let config = Config { accepted_asset, private_fee_recipient };
+        storage.config.initialize(config);
     }
 
+    /// Pays for the tx fee with msg_sender's private balance of accepted asset (AA). The maximum fee a user is willing
+    /// to pay is defined by `max_fee` and is denominated in AA.
+    ///
+    /// ## Overview
+    /// Uses partial notes to implement a refund flow which works as follows:
+    /// 1. `setup_refund` function subtracts the `max_fee` from user's balance of AA, prepares partial notes
+    /// for the `private_fee_recipient` (to obtain the payment in AA for the fee) and for the msg_sender (for refund
+    /// note of the AA) and sets a public teardown function (in which the partial notes will be finalized),
+    /// 2. then the private and public functions of a tx get executed,
```

Comment:
Why does `private_fee_recipient` need to receive the payment privately? The amount is public, and the address is public. Can't we just send them public balance? That'd be much cheaper since there are no logs etc., and also less gates.

Reply:
Makes total sense. Will do the change 👍

Reply:
Maybe not as part of this PR, but I do think we need an example for how someone can pay a fee privately with the following kept private:
- The user
- The FPC contract address
- The asset used to pay the fee.

If there are to be many FPCs on Aztec, then this information _will_ drastically reduce Aztec's global privacy set. 

Ideally we'd have an example of a way to pay fees without leaking any of this info.

Reply:
Created [an issue](https://github.com/AztecProtocol/aztec-packages/issues/10808) for this.

>If there are to be many FPCs on Aztec, then this information will drastically reduce Aztec's global privacy set.

As we discussed in a scrum, this could be alleviated by us integrating the FPC with the token contract and allowing only for a few of those (e.g. 2 stablecoins and ETH).